src/from_fn.rs

```diff
@@ -0,0 +1,132 @@
+//! Support for constructing arrays using a provided generator function.
+
+use crate::{Array, ArraySize};
+use core::{
+    mem::{self, MaybeUninit},
+    ptr,
+};
+
+/// Construct an array type from the given generator function.
+pub trait FromFn<T>: Sized {
+    /// Create array using the given generator function for each element.
+    fn from_fn<F>(cb: F) -> Self
+    where
+        F: FnMut(usize) -> T;
+
+    /// Create an array using the given generator function for each element, returning any errors
+    /// which are encountered in the given generator.
+    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>;
+}
+
+impl<T, U> FromFn<T> for Array<T, U>
+where
+    U: ArraySize,
+{
+    #[inline]
+    fn from_fn<F>(cb: F) -> Self
+    where
+        F: FnMut(usize) -> T,
+    {
+        Array::from_fn(cb)
+    }
+
+    #[inline]
+    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>,
+    {
+        Array::try_from_fn(cb)
+    }
+}
+
+impl<T, const N: usize> FromFn<T> for [T; N] {
+    #[inline]
+    fn from_fn<F>(cb: F) -> Self
+    where
+        F: FnMut(usize) -> T,
+    {
+        core::array::from_fn(cb)
+    }
+
+    // TODO(tarcieri): use `array::try_from_fn` when stable
+    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>,
+    {
+        // SAFETY: an array of `MaybeUninit`s is always valid.
+        let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
+        try_from_fn_erased(&mut array, cb)?;
+
+        // TODO(tarcieri): use `MaybeUninit::array_assume_init` when stable
+        // SAFETY: if we got here, every element of the array was initialized
+        let res: Self = unsafe { ptr::read(array.as_ptr().cast()) };
+        mem::forget(array);
```

Comment:
Hmm, clippy seems to be suggesting this is unnecessary since it's a `MaybeUninit` array

Reply:
Went ahead and removed it